michbl/assignment3/Dictionary.hpp

```diff
@@ -0,0 +1,128 @@
+#ifndef DICTIONARY_HPP
+#define DICTIONARY_HPP
+
+#include <iostream>
+#include <vector>
+#include <string>
+#include <unordered_map>
+#include <set>
+#include "Grid.hpp"
+
+#define END_OF_WORD -1
+
+using namespace std;
+
+// implementing dictionary as trie
+class Dictionary {
+private:
+    struct Node {
+        char c;
+        unordered_map<char, Node *> children;
+
+        Node(char c) : c(c) {
+            this->children.clear();
+        }
+    };
+
+    Node *root;
+    vector<Node *> nodes; // for memory release
+
+    void wordSearchBT(Grid &grid, int row, int col, set<string> &res, string word) {
+        char c;
+        if (!grid.getChar(row, col, c)) {
+            return;
+        }
+        word += c;
+        if (isWord(word)) {
+            res.insert(word);
+        }
+        if (isPrefix(word)) {
+            Direction dirs[8] = {LEFT, RIGHT, UP, DOWN, TOP_LEFT, TOP_RIGHT, BOTTOM_LEFT, BOTTOM_RIGHT};
+            int nextRow, nextCol;
+            for (Direction dir : dirs) {
+                grid.getNextIdxs(row, col, nextRow, nextCol, dir);
+                wordSearchBT(grid, nextRow, nextCol, res, word);
+            }
+        }
+    }
+
+public:
+    Dictionary() {
+        this->root = new Node(END_OF_WORD);
+        nodes.push_back(this->root);
+    }
+
+    Dictionary(const vector<string> words) {
+        this->root = new Node(END_OF_WORD);
+        nodes.push_back(this->root);
+        for (string word : words) {
+            insertWord(word);
+        }
+    }
+
+    ~Dictionary() {
+        for (auto node : this->nodes) {
+            delete node;
+        }
+    }
+
+    void insertWord(const string &word) {
+        unordered_map<char, Node *> *currentChildren = &(this->root->children);

```

Comment:
Tip:
As far as I know, C++ operator precedence rules let you remove the parenthesis:
`&this->root->children;`
here and below.